tools/oeedger8r/Emitter.ml

```diff
@@ -73,62 +71,49 @@ let conv_array_to_ptr (pd : pdecl) : pdecl =
         (PTPtr (tmp_aty, tmp_pa), tmp_declr)
       else (pt, declr)
 
-(** Note that, for a foreign array type [foo_array_t] we will generate
-    [foo_array_t* ms_field;] in the marshalling data structure to keep
-    the pass-by-address scheme as in the C programming language. *)
-let mk_ms_member_decl (pt : parameter_type) (declr : declarator)
-    (isecall : bool) =
-  (* TODO: Clean this up. *)
-  let aty = get_param_atype pt in
-  let tystr =
-    if is_foreign_array pt then
-      sprintf "/* foreign array of type %s */ void" (get_tystr aty)
-    else get_tystr aty
-  in
-  let ptr = if is_foreign_array pt then "*" else "" in
-  let field = declr.identifier in
-  (* String attribute is available for in/in-out both ecall and ocall.
-     For ocall ,strlen is called in trusted proxy code, so no need to
-     defense it. *)
-  let need_str_len_var (pt : parameter_type) =
-    match pt with
-    | PTVal _ -> false
-    | PTPtr (_, pa) ->
-        if pa.pa_isstr || pa.pa_iswstr then
-          match pa.pa_direction with
-          | PtrInOut | PtrIn -> if isecall then true else false
-          | _ -> false
-        else false
-  in
-  let str_len =
-    if need_str_len_var pt then sprintf "    size_t %s_len;\n" field else ""
+(** ----- End code borrowed and tweaked from {!CodeGen.ml} ----- *)
+
+(* Borrowed from upstream OCaml implementation in
+   https://github.com/ocaml/ocaml/pull/2185 *)
+let filter_map f =
```

Comment:
TODO: I need to file a report with OSPO and a proper 3rdparty notice in order to redistribute this code. (Alternatively, we could try to pull in the [Base standard library for OCaml](https://github.com/janestreet/base) which would give me a much more complete FP library 😉)

Reply:
There's a lot more out of Base that would be nice to use though 😄